[feat](streaming job) Introduce streaming job for incremental load

sollhui opened the pull request:
### What problem does this PR solve?

closed https://github.com/apache/doris/issues/56191

### Release note

None

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [ ] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

sollhui committed:
[feat](streaming job) introduce streaming job schedule task (#55790)

### What problem does this PR solve?

introduce streaming job schedule task

JNSimba committed:
[Feature](WIP) Add Create StreamingJob and Alter Job (#55814)

### What problem does this PR solve?

1. Add Create StreamingJob and Alter Job
2. add job and task tvf schema
3. add offset api

sollhui committed:
[feat](streaming job) introduce streaming task scheduler (#55857)

### What problem does this PR solve?

 Introduce streaming task scheduler to schedule all streaming tasks.

JNSimba committed:
[Feature](WIP) add StreamingInsertTask and improve StreamInsertJob (#55862)

### What problem does this PR solve?

1. add StreamingInsertTask For StreamJob
2. Improve StreamInsertJob
3. add insertcommand rewrite tvf params

sollhui committed:
[feat](streaming job) implement offset persistence and replay (#55918)

### What problem does this PR solve?

Implement offset persistence and replay logic(shared noting mode).

JNSimba committed:
[Feature](WIP) add S3 Stream job split offset (#55927)

### What problem does this PR solve?

1. add S3 Stream job split offset
2. fix stream job create bug

JNSimba committed:
[Feature](WIP) Fix streaming job problem (#55949)

### What problem does this PR solve?

Fix streaming job problem

JNSimba committed:
[Feature](WIP) Add fetch meta and fix rewrite tvf problem (#55986)

### What problem does this PR solve?

 Add fetch meta and fix rewrite tvf problem

sollhui committed:
[feat](streaming job) implement offset persistence and replay in cloud mode (#55975)

### What problem does this PR solve?

Implement offset persistence and replay in cloud mode.

sollhui committed:
[fix](streaming job) register listener id when begin transaction (#56056)

### What problem does this PR solve?

Register listener id when begin transaction to ensure before/after
commit logic would be executed.

JNSimba committed:
[Feature](WIP) Add create job case and fix job bug (#56119)

### What problem does this PR solve?

 Add create job case and fix job bug

sollhui committed:
[fix](streaming job) add task commit check and job event lock to ensure exactly-once semantics (#56135)

### What problem does this PR solve?

Add task commit check and job event lock to ensure exactly-once
semantics.

sollhui committed:
[fix](streaming job) fix compile error (#56141)

### What problem does this PR solve?

Fix compile error

sollhui committed:
[fix](streaming job) fix register callback id invalid (#56142)

### What problem does this PR solve?

Fix register callback id invalid.

sollhui committed:
[revert](streaming job) revert "implement offset persistence and replay in cloud mode" (#56149)

### What problem does this PR solve?

Revert "implement offset persistence and replay in cloud mode"

JNSimba committed:
Revert "[revert](streaming job) revert "implement offset persistence and replay in cloud mode"" (#56156)

Reverts apache/doris#56149

JNSimba committed:
[Feature](WIP) Fix Alter Job and schedule bug etc (#56166)

### What problem does this PR solve?

Fix Alter Job and schedule bug etc

Thearas commented:
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

sollhui commented:
run buildall

dataroaring commented on cloud/src/meta-service/meta_service_txn.cpp:
```diff
@@ -610,6 +610,75 @@ void put_routine_load_progress(MetaServiceCode& code, std::string& msg,
+    int64_t txn_id = request->txn_id();
+    if (!request->has_commit_attachment()) {
+        ss << "failed to get commit attachment from req, db_id=" << db_id << " txn_id=" << txn_id;
+        msg = ss.str();
```
code is not set. we can define a marco setting both msg and code.

dataroaring commented on cloud/src/meta-service/meta_service_txn.cpp:
```diff
@@ -1572,6 +1699,11 @@ void MetaServiceImpl::commit_txn_immediately(
+        if (txn_info.load_job_source_type() ==
+            LoadJobSourceTypePB::LOAD_JOB_SRC_TYPE_STREAMING_JOB) {
+            put_streaming_job_meta(code, msg, instance_id, request, txn.get(), db_id);
+        }
```
Should check code and abort.

dataroaring commented on fe/fe-core/src/main/java/org/apache/doris/datasource/property/storage/StorageProperties.java:
```diff
@@ -181,7 +179,9 @@ public static StorageProperties createPrimary(Map<String, String> origProps) {
+                            || LocalProperties.guessIsMe(props)) ? new LocalProperties(props) : null,
+                    props -> (isFsSupport(props, FS_HDFS_SUPPORT)
+                            || HdfsProperties.guessIsMe(props)) ? new HdfsProperties(props) : null
             );
```
Why?

JNSimba replied:
alreayd update

dataroaring commented on fe/fe-core/src/main/java/org/apache/doris/job/common/JobStatus.java:
```diff
@@ -19,7 +19,10 @@
+    /**
+     * For streaming job, if the task has not been created, the job will be in pending state.
+     */
+    PENDING,
```
I am not sure if there a compability issues add an enum at the front.

sollhui replied:
There is no compability issues

dataroaring commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/CreateJobCommand.java:
```diff
@@ -56,10 +58,20 @@ public CreateJobCommand(CreateJobInfo jobInfo) {
+        if (createJobInfo.streamingJob()) {
+            int streamingJobCnt = Env.getCurrentEnv().getJobManager().getStreamingJobCnt();
+            if (streamingJobCnt >= Config.max_streaming_job_num) {
+                throw new JobException("Exceed max streaming job num limit " + Config.max_streaming_job_num);
```
Please hint users how to enlarge.

dataroaring commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/AlterJobCommand.java:
```diff
@@ -0,0 +1,141 @@
+        if (job instanceof StreamingInsertJob) {
+            StreamingInsertJob originJob = (StreamingInsertJob) job;
+            String updateSQL = StringUtils.isEmpty(sql) ? originJob.getExecuteSql() : sql;
+            Map<String, String> updateProps =
```
Users have to provide all propeerties in alter command?

JNSimba replied:
Already fix

dataroaring commented on fe/fe-core/src/main/java/org/apache/doris/job/task/AbstractTask.java:
```diff
@@ -60,7 +60,7 @@ public AbstractTask() {
-    private static long getNextTaskId() {
+    public static long getNextTaskId() {
         // do not use Env.getNextId(), just generate id without logging
         return System.nanoTime() + RandomUtils.nextInt();
```
Is it unique definitely?

dataroaring commented on fe/fe-core/src/main/java/org/apache/doris/job/extensions/insert/InsertTask.java:
```diff
@@ -68,6 +68,8 @@ public class InsertTask extends AbstractTask {
             new Column("LoadStatistic", ScalarType.createStringType()),
             new Column("User", ScalarType.createStringType()),
+            new Column("Offset", ScalarType.createStringType()),
+            new Column("OtherMsg", ScalarType.createStringType()),
```
How to explain OtherMsg to users?

JNSimba replied:
already remove

JNSimba commented on fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:
```diff
@@ -100,16 +100,19 @@ materializedViewStatement
+       commentSpec?
+       DO supportedDmlStatement                                                                                                             #createScheduledJob
    | PAUSE JOB WHERE (jobNameKey=identifier) EQ (jobNameValue=STRING_LITERAL)                                                                #pauseJob
+   | ALTER JOB FOR (jobName=multipartIdentifier) (propertyClause | supportedDmlStatement | propertyClause  supportedDmlStatement)                  #alterJob
```
Here is a reference to ALTER ROUTINE LOAD FOR name

JNSimba commented on fe/fe-core/src/main/java/org/apache/doris/fs/obj/S3ObjStorage.java:
```diff
@@ -630,6 +631,155 @@ public Status globList(String remotePath, List<RemoteFile> result, boolean fileN
+     * @return The largest file name after listObject this time
+     */
+    public String globListWithLimit(String remotePath, List<RemoteFile> result, String startFile,
+            long fileSizeLimit, long fileNumLimit) {
```
Because there is a limit here, and we need to return the largest file in the current list, we wrote a separate method

JNSimba replied:
already reuse

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `100% (0/0)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/56175_7377e32ead526438175862592096952289ca71b5_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/56175_7377e32ead526438175862592096952289ca71b5_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 71.04% (23314/32819) |
| Line Coverage     | 57.31% (242821/423693) |
| Region Coverage   | 52.72% (202274/383644) |
| Branch Coverage   | 54.35% (87008/160095) |

JNSimba committed:
Merge branch 'master' into streaming-job-dev

# Conflicts:
#	fe/fe-core/src/main/java/org/apache/doris/cloud/rpc/MetaServiceClient.java
#	fe/fe-core/src/main/java/org/apache/doris/cloud/rpc/MetaServiceProxy.java
#	fe/fe-core/src/main/java/org/apache/doris/job/extensions/insert/InsertJob.java
#	fe/fe-core/src/main/java/org/apache/doris/job/extensions/insert/InsertTask.java

JNSimba committed:
[Featuren](WIP) add alter job op and fix bug (#56194)

### What problem does this PR solve?

add alter job op and fix bug

sollhui commented:
run buildall

JNSimba committed:
Merge branch 'master' into streaming-job-dev

sollhui commented on fe/fe-core/src/main/java/org/apache/doris/load/loadv2/LoadStatistic.java:
```diff
@@ -123,6 +123,14 @@ public Map<String, String> getCounters() {
+        return fileNum;
+    }
+
+    public long getTotalFileSizeB() {
```
fe/fe-core/src/main/java/org/apache/doris/job/extensions/insert/streaming/StreamingInsertJob.java

sollhui committed:
[fix](streaming job) fix some streaming job bug (#56221)

### What problem does this PR solve?

Fix some streaming job bug.

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/1fb95b2dd116fb5e811597e9cb61f94dce8e3203_1fb95b2dd116fb5e811597e9cb61f94dce8e3203/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/1fb95b2dd116fb5e811597e9cb61f94dce8e3203_1fb95b2dd116fb5e811597e9cb61f94dce8e3203/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 52.35% (17600/33617) |
| Line Coverage     | 37.56% (159791/425438) |
| Region Coverage   | 32.08% (121630/379107) |
| Branch Coverage   | 33.44% (53378/159624) |

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `100% (0/0)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/56175_1fb95b2dd116fb5e811597e9cb61f94dce8e3203_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/56175_1fb95b2dd116fb5e811597e9cb61f94dce8e3203_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 71.07% (23351/32856) |
| Line Coverage     | 57.38% (243420/424252) |
| Region Coverage   | 52.81% (202817/384032) |
| Branch Coverage   | 54.36% (87028/160091) |

JNSimba committed:
[Fix](job) fix job bug (#56225)

### What problem does this PR solve?

[Fix](job) fix job bug

sollhui commented:
run buildall

JNSimba commented on fe/fe-core/src/main/java/org/apache/doris/job/offset/s3/S3Offset.java:
```diff
@@ -0,0 +1,58 @@
+    String startFile;
+    @SerializedName("ef")
+    String endFile;
+    String fileLists;
```
already add

JNSimba commented on fe/fe-core/src/main/java/org/apache/doris/job/extensions/insert/InsertJob.java:
```diff
@@ -96,9 +98,15 @@ public class InsertJob extends AbstractJob<InsertTask, Map<Object, Object>> impl
+            // only execute type = streaming need record
+            .add(new Column("Properties", ScalarType.createStringType()))
+            .add(new Column("Progress", ScalarType.createStringType()))
+            .add(new Column("RemoteOffset", ScalarType.createStringType()))
```
already rename to maxOffset

JNSimba commented on fe/fe-core/src/main/java/org/apache/doris/fs/remote/RemoteFile.java:
```diff
@@ -75,6 +77,22 @@ public void setPath(Path path) {
+    }
+
+    public void setParentPath(String parentPath) {
+        this.parentPath = parentPath;
```
https://github.com/apache/doris/blob/71a8baca13a7d562d8ab7faf93f137302889ec5f/fe/fe-core/src/main/java/org/apache/doris/fs/obj/S3ObjStorage.java#L646-L647

dataroaring commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/CreateJobInfo.java:
```diff
@@ -234,6 +262,32 @@ private AbstractJob analyzeAndCreateJob(String sql, String currentDbName,
+                throw new AnalysisException(e.getMessage());
+            }
+        } else {
+            throw new AnalysisException("Not support this sql : " + sql + " Command class is "
+                    + logicalPlan.getClass().getName() + ".");
```
Only xxx is supported to use with streaming job together.

liaoxin01 commented on fe/fe-core/src/main/java/org/apache/doris/fs/obj/S3ObjStorage.java:
```diff
@@ -528,12 +529,50 @@ ListObjectsV2Response listObjectsV2(ListObjectsV2Request request) throws UserExc
+     * <p>
+     * Copy from `AzureObjStorage.GlobList`
+     */
+    private GlobListResult globListInternal(String remotePath, List<RemoteFile> result, boolean fileNameOnly,
```
Are there any restrictions on the path of the job URI? For example, can paths like `s3://bucket/*/ab/*` and ` s3://bucket/abc.csv` be used?

JNSimba replied:
There is no limit. The partition will be based on the last public path. For example, s3://bucket/*/ab/* will become s3://bucket/{xx,xx}

liaoxin01 commented on fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:
```diff
@@ -4761,7 +4761,7 @@ public void readFromJson(String json) throws IOException {
                         break;
                     case "double":
-                        field.set(this, root.get(attr.name()));
+                        field.set(this, Double.valueOf(root.get(attr.name()).toString()));
```
why change this?

JNSimba replied:
root.get(attr.name()) This is a double string. If it is not forced to be converted, an error will be reported. The int type has been forced to be converted before.

liaoxin01 commented on fe/fe-core/src/main/java/org/apache/doris/job/manager/JobManager.java:
```diff
@@ -328,6 +373,18 @@ public void replayUpdateJob(T job) {
                 .add("msg", "replay update scheduler job").build());
     }
 
+    public void replayUpdateStreamingJob(AlterStreamingJobOperationLog log) {
```
Can the update operations for streaming jobs and other jobs be unified without adding a new AlterStreamingJobOperationLog?

JNSimba replied:
Because other jobs are currently played back globally, and the offset and other information of streaming are stored in txn, the status of playback modification will overwrite the offset and other information.

gavinchou commented on cloud/src/meta-store/keys.cpp:
```diff
@@ -464,6 +466,13 @@ void rl_job_progress_key_info(const RLJobProgressKeyInfo& in, std::string* out)
     encode_int64(std::get<2>(in), out);           // job_id
 }
 
+void streaming_job_meta_key_info(const StreamingJobMetaKeyInfo& in, std::string* out) {
```
who is going recycler these kvs?

gavinchou commented on cloud/src/meta-service/meta_service_txn.cpp:
```diff
@@ -678,6 +749,64 @@ void MetaServiceImpl::get_rl_task_commit_attach(::google::protobuf::RpcControlle
     }
 }
 
+void MetaServiceImpl::get_streaming_task_commit_attach(
```
add UT

gavinchou commented on gensrc/proto/cloud.proto:
```diff
@@ -365,10 +366,21 @@ message RoutineLoadJobStatisticPB {
 
+message StreamingTaskCommitAttachmentPB {
+    optional int64 job_id = 1;
+    optional int64 task_id = 2;
```
what is task id? what is the different/relationship between task_id and job_id?

sollhui replied:
The jobid is determined when creating the job, job will create task to load data, every task has its own task id.

gavinchou commented on cloud/src/meta-store/keys.h:
```diff
@@ -218,6 +218,9 @@ using MetaPendingDeleteBitmapInfo = BasicKeyInfo<24 , std::tuple<std::string, in
 using RLJobProgressKeyInfo = BasicKeyInfo<25, std::tuple<std::string, int64_t, int64_t>>;
 
+//                                                      0:instance_id 1:db_id  2:job_id
+using StreamingJobMetaKeyInfo = BasicKeyInfo<52, std::tuple<std::string, int64_t, int64_t>>;
```
what is the difference between  StreamingJobMetaKeyInfo and RLJobProgressKeyInfo?
if they all components are identical we should use RLJobProgressKeyInfo instead.

gavinchou commented on cloud/src/meta-service/meta_service_txn.cpp:
```diff
@@ -610,6 +610,77 @@ void put_routine_load_progress(MetaServiceCode& code, std::string& msg,
+    TxnCommitAttachmentPB txn_commit_attachment = request->commit_attachment();
+    StreamingTaskCommitAttachmentPB commit_attachment =
+            txn_commit_attachment.streaming_task_txn_commit_attachment();
+    int64_t job_id = commit_attachment.job_id();
```
is this job_id the routine load's job id?

gavinchou reviewed (changes requested):
why not reuse the existing persistent info of routine load for storing streaming job?

sollhui commented on gensrc/proto/cloud.proto:
```diff
@@ -365,10 +366,21 @@ message RoutineLoadJobStatisticPB {
+    optional string offset = 3;
+    optional int64 scanned_rows = 4;
+    optional int64 load_bytes = 5;
+    optional int64 file_number = 6;
```
Previous one.

sollhui commented:
> why not reuse the existing persistent info of routine load for storing streaming job?

There are some differences in metadata information, and we can reuse some logic through pre-refactor or list a to-do list.

sollhui committed:
[fix](streaming job) refactor some logic and add ut (#56292)

### What problem does this PR solve?

Refactor some logic and add ut.

sollhui commented:
run buildall

sollhui commented:
run buildall

sollhui commented:
run buildall

JNSimba committed:
[Feature](wip) improve streamjob and delete alterstreamjob op (#56308)

### What problem does this PR solve?

 improve streamjob and delete alterstreamjob op

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `  6.58% (69/1049)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/56175_813ab6cf14966b03b65cd70373312b8b50b470b3/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/56175_813ab6cf14966b03b65cd70373312b8b50b470b3/fe_report/index.html)

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage `68.57% (96/140)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/f84cc26a5bbc237ba9b8424bc4da1a1f605c9076_f84cc26a5bbc237ba9b8424bc4da1a1f605c9076_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/f84cc26a5bbc237ba9b8424bc4da1a1f605c9076_f84cc26a5bbc237ba9b8424bc4da1a1f605c9076_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 83.84% (1546/1844) |
| Line Coverage     | 67.85% (27717/40853) |
| Region Coverage   | 68.23% (13643/19995) |
| Branch Coverage   | 58.54% (7283/12442) |

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `  6.58% (69/1049)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/56175_f84cc26a5bbc237ba9b8424bc4da1a1f605c9076/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/56175_f84cc26a5bbc237ba9b8424bc4da1a1f605c9076/fe_report/index.html)

sollhui committed:
[fix](streaming job) fix streaming job ut (#56314)

### What problem does this PR solve?

[fix](streaming job) fix streaming job ut

sollhui committed:
fix format

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `  5.21% (52/999)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/56175_32e243b3f500bd7b9de2aac18272092f24b3d98a/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/56175_32e243b3f500bd7b9de2aac18272092f24b3d98a/fe_report/index.html)

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `100% (0/0)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/56175_32e243b3f500bd7b9de2aac18272092f24b3d98a_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/56175_32e243b3f500bd7b9de2aac18272092f24b3d98a_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 71.14% (23386/32875) |
| Line Coverage     | 57.53% (244183/424431) |
| Region Coverage   | 52.99% (203642/384302) |
| Branch Coverage   | 54.52% (87361/160241) |

sollhui commented:
run buildall

hello-stephen commented:
# FE Regression Coverage Report
Increment line coverage `  61.93% (615/993)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/56175_35398e1aff46a80e9a147c6c4f1ca13193717975_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/56175_35398e1aff46a80e9a147c6c4f1ca13193717975_merge_fe/report/index.html)

JNSimba committed:
Merge branch 'master' into streaming-job-dev

# Conflicts:
#	fe/fe-core/src/main/java/org/apache/doris/datasource/property/storage/HdfsPropertiesUtils.java

gavinchou reviewed (commented):
LGTM

JNSimba committed:
[Fix](job) improve job api (#56352)

### What problem does this PR solve?

improve job api

dataroaring reviewed (approved):
LGTM
